### PR TITLE
feat: add C implementation for `math/base/special/ellipe`

### LINKED ISSUE #772: [RFC]: Add C implementation for `@stdlib/math/base/special/ellipe`
### Description

This RFC proposes adding C implementation for [`@stdlib/math/base/special/ellipe`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/ellipe).

```c
double stdlib_base_ellipe( const double m );
```

### Related Issues

#766.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by gunjjoshi**
Resolves #772 

## Description

Added C implementation for @stdlib/math/base/special/ellipe.

> What is the purpose of this pull request?

This pull request:

-   adds c implementation for @stdlib/math/base/special/ellipe

## Related Issues

> Does this pull request have any related issues?

This pull request:

- resolves #772
- a part of #649  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   ☑️ Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by gunjjoshi** [comment]
This is ready for review now. @Pranavchiku

File: lib/node_modules/@stdlib/math/base/special/ellipe/README.md [hunk 1]
```diff
@@ -114,6 +114,96 @@ for ( i = 0; i < 100; i++ ) {
...
+
+### Usage
+
+```c
+#include "stdlib/math/base/special/ellipe.h"
+```
+
+#### stdlib_base_ellipe( m )
+
+Computes the [complete elliptic integral of the first kind][elliptic-integral].
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/README.md]:
```suggestion
Computes the [complete elliptic integral of the second kind][elliptic-integral].
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/include/stdlib/math/base/special/ellipe.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+#ifndef STDLIB_MATH_BASE_SPECIAL_ELLIPE_H
+#define STDLIB_MATH_BASE_SPECIAL_ELLIPE_H
+
+/*
+ * If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+ */
+#ifdef __cplusplus
+extern "C"
+{
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/include/stdlib/math/base/special/ellipe.h]:
```suggestion
extern "C" {
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/include/stdlib/math/base/special/ellipe.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#endif
+
+    /**
+     * Computes the complete elliptic integral of the second kind.
+     */
+    double stdlib_base_ellipe( const double m );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/include/stdlib/math/base/special/ellipe.h]:
```suggestion
/**
* Computes the complete elliptic integral of the second kind.
*/
double stdlib_base_ellipe( const double m );
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/manifest.json [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        "libpath": [],
+        "dependencies": [
+          "@stdlib/constants/float64/half-pi",
+          "@stdlib/constants/float64/pinf",
+          "@stdlib/math/base/special/sqrt",
+          "@stdlib/math/base/special/ln"
+        ]
+      }
+    ]
+  }
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/manifest.json]: incorrect indentation

File: lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+*/
+
+#include "stdlib/math/base/special/ellipe.h"
+#include "stdlib/math/base/special/sqrt.h"
+#include "stdlib/math/base/special/ln.h"
+#include "stdlib/constants/float64/pinf.h"
+#include "stdlib/constants/float64/half_pi.h"
+#include <stdint.h>
+
+static const double HALF_PI = 1.5707963267948966;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c]:
```suggestion
#include <stdint.h>
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+* @param x    value at which to evaluate the polynomial
+* @return     evaluated polynomial
+*/
+static double poly_p12( const double x ) {
+	return 1.5910034537907922 + (x * (0.41600074399178694 + (x * (0.24579151426410342 + (x * (0.17948148291490615 + (x * (0.14455605708755515 + (x * (0.12320099331242772 + (x * (0.10893881157429353 + (x * (0.09885340987159291 + (x * (0.09143962920174975 + (x * (0.0858425915954139 + (x * 0.08154111871830322)))))))))))))))))));
+}
+
+// END: poly_p12
+
+/* End auto-generated functions. */
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c:299]: I am assuming these are not changed generated.

File: lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+
+/**
+* Computes the complete elliptic integral of the second kind.
+*
+* @param x    input value
+* @return     output value
+*
+* @example
+* double out = stdlib_base_ellipe( 0.5 );
+* // returns ~1.3506
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c]:
```suggestion
* double out = stdlib_base_ellipe( 0.5 );
* // returns ~1.351
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+    } else if ( x < 0.9 ) {
+        t = poly_p10( x - 0.875 );
+    } else {
+        td = 0.95 - x;
+        kdm = poly_p11( td );
+        edm = poly_p12( td );
+        km = ellipe( x );
+
+        // To avoid precision loss near 1, we use Eq. 33 from Fukushima (2009):
+        t = ( HALF_PI + ( km * (kdm - edm) ) ) / kdm;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c]:
```suggestion
        t = ( STDLIB_CONSTANT_FLOAT64_HALF_PI + ( km * (kdm - edm) ) ) / kdm;
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+* @return     output value
+*
+* @example
+* double out = stdlib_base_ellipe( 0.5 );
+* // returns ~1.3506
+*/
+double stdlib_base_ellipe( const double m ) {
+    int8_t FLG;
+    double edm;
+    double kdm;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c]:
```suggestion
    double kdm;
    double edm;
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+var abs = require( '@stdlib/math/base/special/abs' );
+var tryRequire = require( '@stdlib/utils/try-require' );
+
+
+// FIXTURES //
+
+var mediumPositive = require( './fixtures/cpp/medium_positive.json' );
+var closeToUnity = require( './fixtures/cpp/close_to_unity.json' );
+var mediumNegative = require( './fixtures/cpp/medium_negative.json' );
+var largeNegative = require( './fixtures/cpp/large_negative.json' );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js]: I appreciate your checks on `mediumNegative` and `largeNegative`.

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
This is good, thanks for your work @gunjjoshi! Please apply requested changes and this will be ready to merge.

File: lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c (hunk 6 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c:299]: Yes.

File: lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js (hunk 10 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js]: Thanks for this and your suggestions, I've made the suggested changes.

**Comment by gunjjoshi** [comment]
@kgryte @Planeshifter @Pranavchiku Any other changes to be made ?

File: lib/node_modules/@stdlib/math/base/special/ellipe/README.md [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 @license Apache-2.0
 
-Copyright (c) 2019 The Stdlib Authors.
+Copyright (c) 2024 The Stdlib Authors.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/README.md]:
```suggestion
Copyright (c) 2019 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/scripts/evalpoly.js [hunk 12]
```diff
@@ -1,7 +1,7 @@
 /**
 * @license Apache-2.0
 *
-* Copyright (c) 2019 The Stdlib Authors.
+* Copyright (c) 2024 The Stdlib Authors.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/ellipe/scripts/evalpoly.js]:
```suggestion
* Copyright (c) 2019 The Stdlib Authors.
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Looks good, one minor comment to add new line at end of file wherever that is missing and this will be good to merge.

File: lib/node_modules/@stdlib/math/base/special/ellipe/README.md (hunk 11 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/ellipe/README.md]: Had a doubt. Should we change the years of already existing files like this one too ?

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/ellipe/README.md]: Reference : https://github.com/stdlib-js/stdlib/pull/1667#issuecomment-1975171823

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/ellipe/README.md]: We change copyright year only for new files, here is detailed reason: https://github.com/stdlib-js/stdlib/pull/1667#issuecomment-1975171823

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/ellipe/README.md]: Yes, so this README.md is not new, it was pre-existent.

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/ellipe/README.md]: Also, I've added the newlines, thanks !

**Review by Pranavchiku — APPROVED** [review]
LGTM. Thanks @gunjjoshi !

File: lib/node_modules/@stdlib/math/base/special/ellipe/examples/c/example.c [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/ellipe.h"
+#include <stdlib.h>
+#include <stdio.h>
+
+int main() {
+    double m;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/ellipe/examples/c/example.c]: This file should be TAB indentation.

File: lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+        FLG += 1;
+    }
+    if ( x == 0.0 ) {
+        return STDLIB_CONSTANT_FLOAT64_HALF_PI;
+    }
+    if ( x == 1.0 ) {
+        return 1.0;
+    }
+    if ( x > 1.0 ) {
+        return 0.0 / 0.0; //NaN
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c]:
```suggestion
        return 0.0 / 0.0; // NaN
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+*
+* @param x    input value
+* @return     output value
+*
+* @example
+* double out = stdlib_base_ellipe( 0.5 );
+* // returns ~1.351
+*/
+double stdlib_base_ellipe( const double m ) {
+    int8_t FLG;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/ellipe/src/main.c]: This file should be TAB indented.

File: lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+	expected = largeNegative.expected;
+	x = largeNegative.x;
+	for ( i = 0; i < x.length; i++ ) {
+		y = ellipe( x[i] );
+		if ( y === expected[i] ) {
+			t.equal( y, expected[i], 'x: '+x[i]+', y: '+y+', expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = 1e9 * EPS * abs( expected[i] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js]: Whoa. Was this tolerance the same as in the JS version? This is massive.

**Review by kgryte — CHANGES_REQUESTED** [review]
@gunjjoshi Thanks for working on this. We need to fix the indentation in a few files.

File: lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js (hunk 16 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js]: You're right, it was a mistake, I forgot to modify certain values from similar package from which I took reference. I have changed it now.

**Comment by gunjjoshi** [comment]
> @gunjjoshi Thanks for working on this. We need to fix the indentation in a few files.

I have made the changes suggested by you, along with fixing the indentation.

File: lib/node_modules/@stdlib/math/base/special/ellipe/examples/c/example.c [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/ellipe.h"
+#include <stdlib.h>
+#include <stdio.h>
+
+int main() {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/ellipe/examples/c/example.c]:
```suggestion
int main( void ) {
```

File: lib/node_modules/@stdlib/math/base/special/ellipe/test/fixtures/cpp/large_negative.json
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/ellipe/test/fixtures/cpp/large_negative.json:1]: Why are these fixtures being added? And if they are needed, why isn't the fixture runner updated?

File: lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+var NINF = require( '@stdlib/constants/float64/ninf' );
+var HALF_PI = require( '@stdlib/constants/float64/half-pi' );
+var EPS = require( '@stdlib/constants/float64/eps' );
+var abs = require( '@stdlib/math/base/special/abs' );
+var tryRequire = require( '@stdlib/utils/try-require' );
+
+
+// FIXTURES //
+
+var mediumPositive = require( './fixtures/cpp/medium_positive.json' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js:36]: You added fixtures, but are not using them in tests. Is there a reason for this?

**Review by kgryte — CHANGES_REQUESTED** [review]
Thanks for the updates, @gunjjoshi. Left a few more comments.

File: lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js (hunk 18 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/ellipe/test/test.native.js:36]: I need to remove the extra fixtures. They had came along with another package's C implementation, from which I took reference. I will delete them in the next commit.

**Review by kgryte — APPROVED** [review]
Okay, this now LGTM. Thanks, @gunjjoshi!

**Comment by kgryte** [comment]
Just an FYI: I had to push a patch due to bugs introduced in this PR: https://github.com/stdlib-js/stdlib/commit/89da7ae0f39326224d5a014ecf537c19f9c4c7c6

For future PRs, be very careful about copying and pasting.

**Comment by gunjjoshi** [comment]
> Just an FYI: I had to push a patch due to bugs introduced in this PR: [89da7ae](https://github.com/stdlib-js/stdlib/commit/89da7ae0f39326224d5a014ecf537c19f9c4c7c6)
> 
> For future PRs, be very careful about copying and pasting.

Definitely I'll keep that in mind. Apologies for this one. Thanks for correcting this.

**Pull request merged**